test: add comprehensive operator lifecycle integration test

diegomrsantos opened the pull request:
## Issue Addressed

Missing test coverage for operator lifecycle management in SSV protocol.

## Proposed Changes

- Add comprehensive operator lifecycle integration test
- Test validates soft delete -> hard delete behavior when operators are removed
- Ensures database integrity through SQL triggers and foreign key constraints

## Additional Info

Critical for validating that operator removal maintains cluster data consistency while properly cleaning up unused operator records.

diegomrsantos committed:
test: add comprehensive operator lifecycle integration test

Implements complete operator lifecycle testing for SSV protocol including:
- Soft delete behavior when operators are removed while still in clusters
- Hard delete via SQL trigger when operators have no cluster references
- Proper database abstraction using existing test utilities
- Code deduplication by extracting shared test helpers to common module

The test validates the critical soft delete -> hard delete flow that maintains
database integrity through foreign key constraints and automated cleanup triggers.

diegomrsantos committed:
fix: address copilot feedback on operator lifecycle test comments

- Fix inaccurate signature verification comment description
- Improve comment formatting consistency

diegomrsantos committed:
refactor: remove redundant memory assertion in verify_operator_stored

The function already verifies memory accessibility by successfully retrieving
the operator from memory state. The additional exists_in_memory assertion
was redundant since it performs the same get_operator call internally.

diegomrsantos committed:
refactor: eliminate duplication in event log creation helpers

Add helper functions to reduce code duplication in create_*_log functions:
- encode_operator_id_topic(): handles operator ID encoding to 32-byte topic
- encode_owner_topic(): handles owner address encoding to 32-byte topic

This eliminates repetitive byte array creation and copying logic across
all event log creation functions, improving maintainability and reducing
the chance of encoding errors.

diegomrsantos committed:
refactor: replace raw SQL with proper database API in tests

Remove database implementation exposure from tests by:
- Replacing raw SQL metadata query with processor.db.state().get_last_processed_block()
- Removing get_metadata() function that exposed internal database schema
- Maintaining proper separation between database API and implementation details

This ensures tests use the public database interface rather than depending on
internal SQL schema, improving maintainability and encapsulation.

diegomrsantos committed:
test: extend operator lifecycle test to cover multi-cluster scenario

Extends the operator lifecycle test to validate behavior across multiple
clusters, ensuring operators are properly soft deleted when removed while
still referenced by clusters, and hard deleted only when no clusters
reference them anymore.

The test now covers:
- Operator creation and participation in two separate clusters
- Operator removal (soft delete from memory, database record remains)
- Sequential cluster removal with proper cleanup verification
- Other operators remain unaffected throughout the lifecycle

diegomrsantos committed:
refactor: replace hardcoded shares with dynamic BLS signature generation

Replace hardcoded VALID_SHARES_DATA constant with dynamic signature generation
to improve test maintainability and eliminate clippy warnings.

- Remove 1312-byte hardcoded hex constant VALID_SHARES_DATA
- Add create_valid_shares_data_for_owner_and_nonce() function for dynamic generation
- Update both operator lifecycle and integration tests to use dynamic shares
- Fix all clippy warnings related to unused imports
- Add required BLS and cryptographic dependencies to test environment
- Ensure both clusters use different nonces for distinct signatures
- Maintain cryptographic validity with proper BLS signature verification

All tests pass with clean linting results.

diegomrsantos commented on anchor/eth/tests/operator_lifecycle.rs:
```diff
@@ -0,0 +1,149 @@
+    // CRITICAL: Use the exact same values as in VALID_SHARES_DATA for signature verification
+    // The shares data contains a BLS signature over "owner:nonce" that must match exactly
+    let cluster_owner =
+        Address::from_str("0x000000633b68f5d8d3a86593ebb815b4663bcbe0").expect("Invalid address");
```
done

dknopik commented on anchor/eth/tests/common/mod.rs:
```diff
@@ -0,0 +1,424 @@
+use std::sync::Arc;
```
This file is `common/mod.rs` while the other modules are just `integration.rs` and `operator_lifecycle.rs`. Unless there are plans to add submodules soon, I'd suggest moving this to `common.rs`

diegomrsantos replied:
done

dknopik commented on anchor/eth/tests/common.rs:
```diff
@@ -0,0 +1,424 @@
+}
+
+/// Helper function to create an OperatorRemoved event log
+#[allow(dead_code)]
```
Same as copilot noted: remove `#[allow(dead_code)]`

Also several occurrences below.

diegomrsantos replied:
Clippy complains that the fn isn't used

dknopik replied:
Ah, okay. I guess what's happening is that each file in the `tests` folder directly is being individually compiled as integration test crate. And while it is used by `integration` and `operator_lifecycle`, it is not used by itself. 

Now I found this note in the rust docs on integration tests: "Creating the module as tests/common.rs also works, but is not recommended because the test runner will treat the file as a test crate and try to run tests inside it."

I did not know that - this is the first time where I heard about `xyz.rs` being treated differently from `xyz/mod.rs` :D

In that case, your previous approach of `common/mod.rs` was actually more correct. sorry for the back and forth. I believe maybe then the `allow(dead_code)` can be removed.

diegomrsantos replied:
It was also happening with the previous approach :D

diegomrsantos replied:
pushed a fix

diegomrsantos commented on anchor/eth/tests/common.rs:
```diff
@@ -0,0 +1,361 @@
+}
+
+/// Create a KeySplit mode EventProcessor for testing
+#[allow(dead_code)]
```
we need this

dknopik replied:
why? the function appears to be used, the `allow` should not be needed

dknopik commented on anchor/eth/tests/common/mod.rs:
```diff
@@ -0,0 +1,424 @@
+    let tx = conn.transaction().expect("Failed to start transaction");
+
+    assertions::operator::exists_not_in_db(operator_id, &tx);
+}
```
These function do the exact same thing, is this your intention?

diegomrsantos replied:
They should be different, thanks. Fixed.

diegomrsantos committed:
refactor: extract hardcoded test address to constant

Extract hardcoded cluster owner address to TEST_CLUSTER_OWNER constant
in common test module for better maintainability. Update both
operator_lifecycle.rs and integration.rs to use the shared constant.

diegomrsantos committed:
fix: implement proper soft vs hard delete verification

Add database API method is_operator_soft_deleted() to distinguish
between operators that are soft deleted (removed=TRUE in database)
versus hard deleted (completely removed from database).

Update test verification functions to properly check soft delete
vs hard delete states using the new API method instead of relying
on identical behavior that couldn't distinguish the two cases.

diegomrsantos commented on anchor/eth/tests/common/mod.rs:
```diff
@@ -0,0 +1,444 @@
+        .expect("Failed to check if operator exists in database");
+
+    assert!(
+        !is_soft_deleted,
```
I'm probably gonna improve this

diegomrsantos replied:
done

diegomrsantos committed:
refactor: improve operator status checking with enum-based API

Add OperatorStatus enum to clearly represent operator states (Active,
SoftDeleted, NotFound) and refactor database API to use a single
get_operator_status function with focused helper methods.

This simplifies the previous complex get_operator_with_status
implementation and provides cleaner test verification functions.

diegomrsantos committed:
refactor: move common/mod.rs to common.rs

Simplify module structure since there's only one common file.
Follow Rust convention of using mod.rs only when there are
multiple submodules in a directory.

diegomrsantos committed:
refactor: rename operator_doesnt_exist to does_operator_exist

Use positive naming convention to avoid double negatives.
The function now returns true if operator exists (either active
or soft deleted) making the API clearer and test logic simpler.

dknopik commented on anchor/database/src/operator_operations.rs:
```diff
@@ -97,4 +107,46 @@ impl NetworkDatabase {
+        id: OperatorId,
+        tx: &Transaction<'_>,
+    ) -> Result<OperatorStatus, DatabaseError> {
+        let query = "SELECT removed FROM operators WHERE operator_id = ?1";
```
Move this query to be a constant in `sql_operations.rs` to keep them all in one place.

diegomrsantos replied:
done

diegomrsantos committed:
refactor: use idiomatic pattern for shared test utilities

Move tests/common.rs to tests/common/mod.rs following Rust Book guidance.
Replace individual #[allow(dead_code)] with module-level #![allow(dead_code)]
since integration tests are compiled independently.

diegomrsantos committed:
refactor: move GET_OPERATOR_STATUS query to sql_operations constants

dknopik reviewed (approved):
would be good to somehow avoid the `allow(dead_code)`, but we can take a look at that later ;)
